Backends/Wayland/input.lisp

```diff
@@ -0,0 +1,119 @@
+
+(defmethod process-next-event ((port wayland-port) &key wait-function (timeout nil))
+  (let ((*wayland-port* port)
+        (*wayland-wait-func* wait-function))
+    (let ((event (progn (when timeout
+                          (sleep timeout))
+                        (wlc:wl-display-dispatch
+                         (wayland-port-display *wayland-port*)))))
+      (format t "process-next-event status? ~a~%" event)
+      ;; Return t for now since I seem beholden to the wlc event loop and I'm
+      ;; not sure when I should return nil
+      (case event
+        ((nil)
+         (if (maybe-funcall wait-function)
+             (values nil :wait-function)
+             (values nil :timeout)))
+        ((t)
+         (values nil :wait-function))
+        (otherwise
+         (prog1 t
+           (unless (numberp event) ; status numbers are returned for events we
+                                   ; haven't translated yet.
+             (distribute-event port event))))))))
+
+;;; Now we connect all of wayland-client events to our port's events
+
+(defmethod xdg:xdg-surface-configure ((surface xdg:xdg-surface) serial)
+  ;; We assume we handle configuration events immediately, so we can
+  ;; acknowledge right away. TODO: find where in the CLIM flow this should
+  ;; really happen. This ACK is very important for the PROCESS-NEXT-EVENT to
+  ;; continue even though it might be in the wrong place
+  (xdg:xdg-surface-ack-configure surface serial))
+
+(defmethod xdg-surface-configure :after ((surface xdg:xdg-surface) serial)
+  (format *debug-io* "xdg surface configured ~a  surface: ~a~%" serial surface))
+
+(defmethod xdg:xdg-toplevel-configure
+    ((wayland-proxy wayland-xdg-toplevel) width height states)
+  (declare (ignore states wayland-proxy))
+  (format t "toplevel configure event: w: ~a  h: ~a~%" width height)
+  (let* ((sheet (graft *wayland-port*))
+         (clim-event (make-instance 'window-configuration-event
+                                    :width width
+                                    :height height
+                                    :x 0
+                                    :y 0
+                                    :sheet sheet
+                                    :region (make-bounding-rectangle 0 0 width height0))))
+    (distribute-event *wayland-port* clim-event)))
+
+(defmethod xdg:xdg-toplevel-close ((wayland-proxy wayland-xdg-toplevel))
+  (declare (ignore wayland-proxy))
+  (format *debug-io* "close toplevel event ~s~%" *application-frame*)
+  ;; QQQQ this feels terrible but it seems *application-frame* isn't
+  ;; bound. Extremely hacky way to get top-level-frame-sheet
+  (let* ((sheet (frame-top-level-sheet
+                 (first (frame-manager-frames
+                         (find-frame-manager :port *wayland-port*)))))
+         (clim-event (make-instance 'window-manager-delete-event
+                                    :sheet sheet)))
+    ;; (break)
+    (distribute-event *wayland-port* clim-event)))
```

Comment:
doesn't this event indicate in any way which window is closed? if it does, then you may do a reverse lookup to check what sheet is associated with the mirror.